partner/WHMCS/modules/servers/jetpack/jetpack.php

```diff
@@ -0,0 +1,288 @@
+ * and client secret stored when the product was created in
+ * the WHMCS product settings.
+ *
+ *
+ * @param $params
+ * @return mixed
+ * @throws Exception
+ */
+function get_access_token($params)
+{
+
+    $oauthURL = "https://public-api.wordpress.com/oauth2/token";
+
+    $credentials = array (
+        'client_id' => $params['configoption1'],
+        'client_secret' => $params['configoption2'],
+        'grant_type' => 'client_credentials',
+        'scope' => 'jetpack-partner'
+    );
+
+    $response = make_api_request($oauthURL, null, $credentials);
+    if (isset($response->access_token)) {
+        return $response->access_token;
+    } else {
+        return $response;
+    }
+}
+
+
+/**
+ * Make an API request for authenticating and provisioning or
+ * cancelling a Jetpack plan
+ *
+ * @param $url
+ * @param $data
+ * @param string $auth
+ * @return mixed
+ * @throws Exception
+ */
+function make_api_request($url, $auth = null, $data = null)
+{
+    if (isset($auth)) {
+        $auth = "Authorization: Bearer " . $auth;
+    }
+
+    $curl = curl_init();
+    curl_setopt_array($curl, array(
+        CURLOPT_HTTPHEADER => array($auth),
+        CURLOPT_URL => $url,
+        CURLOPT_RETURNTRANSFER => true,
+        CURLOPT_ENCODING => "",
+        CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
+        CURLOPT_POSTFIELDS => $data,
+        CURLOPT_CUSTOMREQUEST => "POST"
+    ));
+
+    $response = curl_exec($curl);
+    $http_code = curl_getinfo($curl, CURLINFO_HTTP_CODE);
+
+    if ($response && $http_code >= 400) {
+        if ($data['client_id']) {
+            //If client_id is set in the data array when this function was called then we are trying to get an access
+            // token and are unsuccessful which likely means the client id or client secret are not valid.
+            return 'JETPACK MODULE: There was a problem getting an access token for your Jetpack hosting partner 
+            account. This usually means the Client Id or Client Secret provided when creating the product are invalid';
+        } elseif ($data['siteurl']) {
+            //If the siteurl is set in the data array we are attempting to provision a plan. There are few errors that
+            //could show up here but its likely the user did not enter siteurl or local_user field correctly.
+            return 'JETPACK MODULE: There was a problem provisioning a Jetpack Plan for ' . $data['siteurl'] . '. The 
+            response from the request was ' . $response . '. This usually means the url provided does not have a valid 
+            WordPress installation or the local_user field provided for provisioning is not valid.';
+        }
+        return 'JETPACK MODULE: There was a problem with provisioning or terminating a Jetpack Plan. 
+        The response was ' . $response;
+    } elseif (curl_error($curl)) {
+        throw new Exception('Unable to connect: ' . curl_errno($curl) . ' - ' . curl_error($curl));
```

Comment:
In what cases do we return the error message as a string and in what cases do we throw an exception?

Reply:
@nb I've refactored this and make_api_request no longer handles parsing the response and will not return any error strings. It either gets a response and returns it or throws an exception if there's a curl_error or the response is empty. Updates on this are also in the P2.